src/http_microhttpd.c

```diff
@@ -412,6 +450,12 @@ static int authenticated(struct MHD_Connection *connection,
 			return authenticate_client(connection, ip_addr, mac, redirect_url, client);
 	} else if (check_authdir_match(url, config->denydir)) {
 		auth_client_action(ip_addr, mac, AUTH_MAKE_DEAUTHENTICATED);
+		if(config->bin_voucher) {
+			char *cmd = NULL;
+
+			safe_asprintf(&cmd,"%s auth_update %s %s %d", config->bin_voucher, mac, client->voucher, time(NULL) - client->added_time);
```

Comment:
This one hurt's me, because client->voucher is filled fprintf() into cmd.

I see you sanitize the input before which is good.
Even then I like to see client->voucher ends up as as list entry of the argument list passed to execv()
I know NDS has some corpes lying around because it's using /bin/sh (even 2 times in a call - WTF).
So can you please decouple execute() a little bit more:
So that execute looks like  `execute(const char *file, const char *argv[], int quiet, char *msg, size_t msg_len)` and it's passing argument `*file` `*argv` direct execv().

If you don't like to do it, please tell me, so I can do it as requirement for your commit.